src/render/color.rs

```diff
@@ -28,13 +48,272 @@ impl Color {
+    /// - "RRGGBB"
+    /// - "AARRGGBB"
+    /// - "#RRGGBB"
+    /// - "#AARRGGBB"
+    /// - "0xRRGGBB"
+    /// - "0xAARRGGBB"
+    pub fn parse(s: String) -> Option<Color> {
+        if s.starts_with("#") {
+            Color::parse(s[1..].to_string())
+        } else if s.starts_with("0x") {
+            Color::parse(s[2..].to_string())
+        } else if s.len() == 8 {
+            Color::parse_argb(s)
+        } else if s.len() == 6 {
+            Color::parse_rgb(s)
+        } else {
+            None
+        }
+    }
+
+    /// Parses an ARGB String into a Color
+    fn parse_argb(s: String) -> Option<Color> {
+        if s.len() == 8 {
+            let mut chars = s.chars();
+            let alpha = Color::parse_color(&mut chars);
+            let color = Color::parse_rgb(chars.as_str().to_string());
+            if WLC_BUG_INVERT_RED_BLUE {
+		// Due to the bug, the colors are already inverted, so in the returned color
+                // red is blue and blue is red.
+                alpha.and_then(|a| color.map(|rgb| Color::rgba(rgb.blue, rgb.green, rgb.red, a)))
+            } else {
+                alpha.and_then(|a| color.map(|rgb| Color::rgba(rgb.red, rgb.green, rgb.blue, a)))
+            }
+        } else {
+            None
+        }
+    }
+
+    /// Parses a RGB String into a Color
+    fn parse_rgb(s: String) -> Option<Color> {
+        if s.len() == 6 {
+            let mut chars = s.chars();
+            let red   = Color::parse_color(&mut chars);
+            let green = Color::parse_color(&mut chars);
+            let blue  = Color::parse_color(&mut chars);
+            red.and_then(|r| green.and_then(|g| blue.map(|b| Color::rgba(r, g, b, 255))))
+        } else {
+            None
+        }
+    }
+
+    /// Parses exactly one single color value from a String (eg "AA", "RR", "GG" or "BB")
+    fn parse_color(chars: &mut Chars) -> Option<u8> {
+        let mut next_two_chars = chars.take(2);
+        let digit1 = Color::parse_next_char(next_two_chars.next());
+        let digit2 = Color::parse_next_char(next_two_chars.next());
+        digit1.and_then(|i1| digit2.map(|i2| (i1 << 4) | i2))
+    }
+
+    fn parse_next_char(c: Option<char>) -> Option<u8> {
+        c.and_then(Color::hex_to_u8)
+    }
+
+    /// Converts a hex char into a u8
+    fn hex_to_u8(c: char) -> Option<u8> {
+        c.to_digit(16).map(|x| (x as u8))
+    }
+
 }
 
 impl From<u32> for Color {
     fn from(val: u32) -> Self {
-        let blue = ((val & 0xff0000) >> 16) as u8;
-        let green = ((val & 0x00ff00) >> 8) as u8;
-        let red = (val & 0x0000ff) as u8;
-        Color::solid_color(red, green, blue)
+        let red   = ((val & 0xff0000) >> 16) as u8;
+        let green = ((val & 0x00ff00) >> 8)  as u8;
+        let blue  =  (val & 0x0000ff)        as u8;
+        Color::rgba(red, green, blue, 255)
+    }
+}
+
+mod test {
+
+    use ::render::Color;
+
+    #[test]
+    fn test_from_u32() {
+        let hex_red   = 0xFF0000;
```

Comment:
use spaces, not tabs.

Reply:
It's fixed on the lines below, this line doesn't seem to have tabs.